### PR TITLE
#160056376 view edit and profile

### PR DESCRIPTION
**Pull request by tutugodfrey**
#### What does this PR do?
- Add functionality to get and update profiles

#### Description of Task to be completed?
- Create Profile component
- Modify view Profile conponent
- Add function to get profile and save to redux store
- Add function for log-in to update their profile
- update localStorage when profile update is successful

#### How should this be manually tested?
N/A

#### Any background context you want to provide?
N/A

#### Screenshots (if appropriate)
image
image
image
image

#### What are the relevant pivotal tracker stories?
#160056376

### CODE REVIEW

File: src/components/Logout.js [hunk 1]
```diff
@@ -1,26 +1,26 @@
-import React from 'react';
+// import React from 'react';
```
Comment (JCanaks) [inline src/components/Logout.js]: instead of commenting put the enter component why not leave a basic component like
```
import React from 'react';

const Logout = () => (
  <div>
    <h2>Logout Component</h2>
  </div>
);
export default Logout;
```

File: src/configs/routes.js [hunk 2]
```diff
@@ -75,7 +75,7 @@ export default [
   },
   {
     id: 9,
-    type: AuthRoute,
+    type: Route,
```
Comment (JCanaks) [inline src/configs/routes.js]: Editing a profile should be a protected route. `PrivateRoute component` should be use not `Route`

File: src/store.js [hunk 3]
```diff
@@ -7,12 +7,13 @@ import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
 import rootReducer from './reducers';
 import errorMiddleware from './middleware/errorMiddleware';
 import rootSaga from './sagas';
+import authMiddleware from './middleware/authMiddleware';
 
 const promiseTypeSuffixes = ['LOADING', 'SUCCESS', 'FAILURE'];
 const sagaMiddleware = createSagaMiddleware();
-
+// createLogger({ collapsed: true }),
```
Comment (JCanaks) [inline src/store.js]: Pls remove commented code

File: src/views/index.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import 'babel-polyfill';
```
Comment (JCanaks) [inline src/views/index.js]: Should this file be within the views folder? We have one already in the source folder

**Comment by megame24** [comment]
Please fix your hound violations

File: package.json [hunk 5]
```diff
@@ -7,7 +7,7 @@
     "build": "webpack --mode production --env.mode=production --open",
     "dev": "NODE_ENV=development webpack-dev-server --mode development --env.mode=development --open --hot",
     "lint": "eslint src",
-    "test": "jest ",
+    "test": "jest /tests/components/GuestRoute.test.js",
```
Comment (megame24) [inline package.json]: revert this change please

File: src/actions/updateProfile.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import axiosInstance from '../services/requestHandler';
+import types from './actionTypes';
+
+const updateProfile = (path, data) => ({
```
Comment (megame24) [inline src/actions/updateProfile.js]: Put all your profile related actions into one `profileAction` file

File: src/components/EditProfile.jsx [hunk 7]
```diff
@@ -0,0 +1,117 @@
+import React, { Component } from 'react';
+import { Form, Button, Container } from 'semantic-ui-react';
+
+export default class EditProfile extends Component {
```
Comment (megame24) [inline src/components/EditProfile.jsx]: Since this component is not handling internal state, you can make it a functional component

File: src/components/LoginForm.js [hunk 8]
```diff
@@ -0,0 +1,156 @@
+import React from 'react';
```
Comment (megame24) [inline src/components/LoginForm.js]: You shouldn't have `LoginForm` changes as your changes, please `cherry pick` your commits. This article can help you do that: https://www.clock.co.uk/insight/deleting-a-git-commit

File: src/store.js [hunk 9]
```diff
@@ -11,7 +11,7 @@ import authMiddleware from './middleware/authMiddleware';
...
```
Comment (megame24) [inline src/store.js]: you can get rid of the commented code

File: src/views/user/EditProfilePage.jsx [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    this.handleUpdateProfile = this.handleUpdateProfile.bind(this);
+    this.state = {};
+  }
+
+
+
+  handleChange(event) {
+    event.persist();
+    const { name, value, id } = event.target;
+    if((name === 'firstName' || name === 'lastName' || name === 'email'|| name === 'bio')
```
Comment (megame24) [inline src/views/user/EditProfilePage.jsx]: Please comment this code to provide more clarity on what you are trying to achieve

File: src/views/user/EditProfilePage.jsx [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    this.setState(state => (
+      {
+        formData: {
+          ...state.formData, [name]: value,
+        }
+      }
+    ));
+  };
+
+  componentWillMount() {
```
Comment (megame24) [inline src/views/user/EditProfilePage.jsx:31]: Why don't you use `componentDidMount`, because `componentWillMount` is deprecated

File: src/views/user/EditProfilePage.jsx [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    this.setState({
+      userRedirect: false,
+      formData: this.props.userProfile,
+    })
+  }
+
+  mouseOutHandler(event) {
+    const { name, value, id } = event.target;
+    let eleClass = event.target.className;
+    if((name === 'firstName' || name === 'lastName' || name === 'email' || name === 'bio')
```
Comment (megame24) [inline src/views/user/EditProfilePage.jsx]: Add comments to provide clarity

**Comment by megame24** [comment]
Please try and achieve a reasonable test coverage for your components.

File: src/views/user/EditProfilePage.jsx (hunk 11 above)
Comment (tutugodfrey) [reply src/views/user/EditProfilePage.jsx:31]: Ok!

File: src/actions/updateProfile.js (hunk 6 above)
Comment (tutugodfrey) [reply src/actions/updateProfile.js]: make sense

File: src/components/EditProfile.jsx [hunk 13]
```diff
@@ -0,0 +1,114 @@
+import React, { Component } from 'react';
+import { Form, Button, Container } from 'semantic-ui-react';
+
+const EditProfile = ({ formData, handleSubmitForm, handleChange, mouseOutHandler, isLoading}) => {
+  console.log(formData, 'ASDASDASDASD')
```
Comment (mppdg) [inline src/components/EditProfile.jsx]: don't forget to remove the `console.log`

**Review by mppdg — COMMENTED** [review]
Please fix the Hound violations. Let me know if you need help

File: src/views/user/ProfilePage.jsx [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+      profile = this.props.userProfile;
+    }
+
+    return (
+        <div id="profile-page">
+        <Profile
+          profile={profile}
+          username={username}
+          owner={this.props.userProfile.username}
+          errors={this.props.errors}
```
[the remark this answers is not in the document]
Comment (tutugodfrey) [reply src/views/user/ProfilePage.jsx:45]: outdated comment

File: src/components/EditProfile.jsx [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+              name="firstName"
+              value={firstName}
+            />
+            <span className="status-box">{''}</span>
+          </div>
+          <div className="form-group">
+            <label htmlFor="lastname">
+              Lastname
+            </label>
+            <input
```
[the remark this answers is not in the document]
Comment (tutugodfrey) [reply src/components/EditProfile.jsx:68]: outdated comment

File: src/views/user/EditProfilePage.jsx [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+          .replace('valid-email', 'invalid-email');
+        targetElement.setAttribute('class', elementClassValue);
+        const spanEle = event.target.nextSibling;
+        spanEle.innerHTML = `please enter a valid ${event.target.name}`;
+        return event;
+      }
+    }
+  }
+
+  /**
```
[the remark this answers is not in the document]
Comment (tutugodfrey) [reply src/views/user/EditProfilePage.jsx]: outdated comment

File: src/views/user/EditProfilePage.jsx [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+        const ele = document.getElementById(id);
+        ele.setAttribute('class', eleClass);
+        const spanEle = event.target.nextSibling;
+        spanEle.innerHTML = `please enter a valid ${event.target.name}`;
+       return;
+      }
+    }
+  }
+
+  /**
```
[the remark this answers is not in the document]
Comment (tutugodfrey) [reply src/views/user/EditProfilePage.jsx]: outdated comment

File: src/sagas/profileSaga.js [hunk 18]
```diff
@@ -25,4 +25,28 @@ export function* saveProfileOnPersistLogin() {
...
+ * @returns {null} null
+ */
+export function* updateUserOnProfileUpdate() {
+  const getUserDetails = yield take(`${types.PROFILE_UPDATE}_SUCCESS`);
+  let userDetails = getUserDetails.payload.data.user;
+  const { username } = userDetails;
+  userDetails = JSON.stringify(userDetails);
+  localStorage.setItem('user', userDetails);
+
+  /* eslint-disable no-restricted-globals */
```
Comment (rhenshaw56) [inline src/sagas/profileSaga.js]: why the eslint disabling ?

File: src/views/user/EditProfilePage.jsx [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+   */
+  constructor() {
+    super();
+    this.handleChange = this.handleChange.bind(this);
+    this.mouseOutHandler = this.mouseOutHandler.bind(this);
+    this.handleUpdateProfile = this.handleUpdateProfile.bind(this);
+    this.state = {};
+  }
+
+  /* eslint-disable class-methods-use-this */
```
Comment (rhenshaw56) [inline src/views/user/EditProfilePage.jsx]: 👀

**Review by rhenshaw56 — APPROVED** [review]

**Comment by rhenshaw56** [comment]
remove the disabling and have on of your team mates go through it so we should gat this in asap

**Review by mppdg — APPROVED** [review]

**Review by Fattylee — APPROVED** [review]

**Review by chidioguejiofor — DISMISSED** [review]
Nice work in completing this feature. 
I tested this branch locally and  found that you were not uploading the user image to any web server instead you were persisting a `fake url` to the database. 
What do you think about first uploading the image to cloudinary and then using the secure_url to create user profile

Also, I would like to know what you think about adding padding to the view profile page below

screen shot 2018-09-26 at 7 10 02 pm

After editing my profile and uploading an image, here is how the page looks

screen shot 2018-09-26 at 7 11 01 pm
